sw/device/lib/crypto/impl/kdf.c

```diff
@@ -28,8 +29,109 @@ otcrypto_status_t otcrypto_kdf_kmac(
     otcrypto_kmac_mode_t kmac_mode, const otcrypto_const_byte_buf_t kdf_label,
     const otcrypto_const_byte_buf_t kdf_context, size_t required_byte_len,
     otcrypto_blinded_key_t *keying_material) {
-  // TODO: Implement KMAC-KDF.
-  return OTCRYPTO_NOT_IMPLEMENTED;
+  // Check NULL pointers.
+  if (key_derivation_key.keyblob == NULL || keying_material == NULL) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check the private key checksum.
+  if (integrity_blinded_key_check(&key_derivation_key) != kHardenedBoolTrue) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // The underlying KMAC implementation is not hardened yet.
+  if (key_derivation_key.config.security_level !=
+      kOtcryptoKeySecurityLevelLow) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
+
+  // Ensure `key_mode` is either KMAC-128 or KMAC-256.
+  if (key_derivation_key.config.key_mode != kOtcryptoKeyModeKdfKmac128 &&
+      key_derivation_key.config.key_mode != kOtcryptoKeyModeKdfKmac256) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check if `key_mode` values match the input arg.
+  if (!((key_derivation_key.config.key_mode == kOtcryptoKeyModeKdfKmac128 &&
+         kmac_mode == kOtcryptoKmacModeKmac128) ||
+        (key_derivation_key.config.key_mode == kOtcryptoKeyModeKdfKmac256 &&
+         kmac_mode == kOtcryptoKmacModeKmac256))) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Validate key length of `key_derivation_key`.
+  if (key_derivation_key.config.hw_backed == kHardenedBoolTrue) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  } else if (key_derivation_key.config.hw_backed == kHardenedBoolFalse) {
+    // Force `key_length` to be larger than minimum supported security level (
+    // 128-bit = 16 bytes).
+    if (key_derivation_key.config.key_length < 128 / 8) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+    if (2 * key_derivation_key.config.key_length !=
+        key_derivation_key.keyblob_length) {
+      return OTCRYPTO_BAD_ARGS;
+    }
```

Comment:
nit: You can eliminate the magic number with `keyblob_num_words`:
```suggestion
    if (key_derivation_key.keyblob_length != keyblob_num_words(key_derivation_key.config) * sizeof(uint32_t)) {
      return OTCRYPTO_BAD_ARGS;
    }
```

Reply:
Because `keyblob_num_words` rounds things up to word sizes without raising erros, I am worried that it makes these checks looser. For example, if the caller uses `key_derivation_key.config.key_length = 15` (instead of 16 bytes), with `keyblob_length = 32`, this should be rejected by KMAC driver. However, as far as I see, using `keyblob_num_words` or `keyblob_share_num_words` truncates this precision.

Reply:
On a second thought, I think these two can be two separate checks. Here, I will use `keyblob_num_words` function for every `key_length` to `keyblob_length` comparison. `key_length % 4 !=0` checks should be handled separately. We can mark this as resolved.